### ISSUE TITLE
[1.10.2-PE1.0.6B] Inventory Tweaks won't sort alchemical chest/bag

### ISSUE THREAD

**Issue by peregrinus11** [sinkillerj/ProjectE#1386]
Inventory Tweaks version 1.61-58

When clicked by middle-button, items are only stacked one after other (empty spaces filled), but container will not be sorted and subsequent sorting actions will return message "Error Failed to sort container: null (l101 in HandlerSorting.java)".

Was [already reported](https://github.com/Inventory-Tweaks/inventory-tweaks/issues/364) on Inventory Tweaks project page and rejected as non-vanilla containers should provide more information.

**Comment by Lilylicious** [comment]
I _suspect_ that it is due to our slots apparently not being able to be swapped. Starting in the 1.9.4 version, any slots that contain an actual item (excluding the philostone crafting grid) will not swap items between the hand and the slot, unless the slot is empty. 

I suspect the sorting method uses the same function. The inventories still combine stacks as they should, it's only the moving of items that fails (but only when trying to swap two items. Empty slots get filled as they should).

**Comment by howtonotwin** [comment]
Related to MinecraftForge/MinecraftForge#3059 perhaps?

**Comment by Lilylicious** [comment]
That does seem like the likely candidate. I'll look into it after uni today.

**Comment by Lilylicious** [comment]
I believe our overriding of isItemValid in our slot classes to only allow EMC items prevents us from benefiting from the isItemValid changes in MinecraftForge/MinecraftForge#3273

**Comment by williewillus** [comment]
I dont think this ever worked even in 1.7

**Comment by Lilylicious** [comment]
I've tested. In 1.7.10 and 1.8.9, you could swap items like in vanilla chests. 1.9.4 and on it doesn't work.

**Comment by Lilylicious** [comment]
Which reminded me to check what branch I'm looking at and I am no longer pretty sure I know how to fix it.

**Comment by Lilylicious** [comment]
Figured it out (I think). The item replacing "hack" was implemented in forge for IItemHandlerModifiable. We're using CapabilityItemHandler. We'll need to implement a similar hack in our own overrides, or change the itemhandler we use.

**Comment by williewillus** [comment]
Oh, I meant that invtweaks sorting never worked in 1.7 iirc

IItemHandlerModifiable is exactly what we're using, Capabilityitemhandler is just the corresponding Cap Object

**Comment by Lilylicious** [comment]
Huh. In that case I still have no idea why we can't replace items in slots. It's entirely possible that the invtweaks thing is unrelated. I'm home now, so time to go bug hunting!

**Comment by williewillus** [comment]
Are we sure that's the issue and that it isn't the support being incompletely implemented?

**Comment by Lilylicious** [comment]
Not at all. I'm updating a test pack to a forge version with the fix, will see if that's fixed, then check if invtweaks suddenly works or not. Thing is, half of the default sorting method works (moving to empty slots and combining stacks), but moving items doesn't work.

**Comment by Lilylicious** [comment]
Confirmed. Given a minecraftForge version later than (I think) 12.18.2.2110 or possibly a few more, both item replacing and invTweaks default sorting works fine. The other sorting modes do not work, but that's not a surprise. Pressing the sort button (middle mouse for me) more than once creates an error, but lots of mods do that. Might look into supporting other sorting modes, but it's not a priority.
